lib/ddtrace/tracer.rb

```diff
@@ -164,17 +166,56 @@ def set_tags(tags)
 
     # Guess context and parent from child_of entry.
     def guess_context_and_parent(child_of)
-      # call_context should not be in this code path, as start_span
-      # should never try and pick an existing context, but only get
-      # it from the parameters passed to it (child_of)
-      return [Datadog::Context.new, nil] unless child_of
+      case child_of
+      when Context
+        [child_of, child_of.current_span]
+      when SpanOperation
+        [child_of.context, child_of]
+      when Span
+        [nil, child_of]
+      else
```

Comment:
Yeah, tend to agree... switch statements like this are a bit of a code smell. Issue is we then have to have all these objects produce this kind of interface, and "know" about context management.

I want to hold off on that kind of change for now; context management will undergo an overhaul, and this might be eliminated in the process anyways. We can always do another refactoring pass if we don't.